pkg/replicate/scheme.go

```diff
@@ -203,9 +204,35 @@ func (rs *replicationScheme) execute(ctx context.Context) error {
 		return availableBlocks[i].BlockMeta.MinTime < availableBlocks[j].BlockMeta.MinTime
 	})
 
+	// Replicate concurrently
+	var wg sync.WaitGroup
+	errChan := make(chan error)
+	finishChan := make(chan struct{})
+
 	for _, b := range availableBlocks {
-		if err := rs.ensureBlockIsReplicated(ctx, b.BlockMeta.ULID); err != nil {
-			return errors.Wrapf(err, "ensure block %v is replicated", b.BlockMeta.ULID.String())
+		wg.Add(1)
+		go func(b *metadata.Meta) {
+			defer wg.Done()
+			if err := rs.ensureBlockIsReplicated(ctx, b.BlockMeta.ULID); err != nil {
```

Comment:
* Set number of workers and allow the user to configure it (bounded concurrency).
* I think we need another mode that group those by external labels and parallelize only between external labels? under flag (full concurrency vs external labels grouping)

Reply:
Thanks for the coemmnt. Working on these.

Reply:
Hello @bwplotka
What do the external labels mean here?  What level should we group by the blocks ？Should we group the blocks by any distinct label (k-v pairs) or label name?

Reply:
Some ideas:

A) We could use grouper directly: https://github.com/thanos-io/thanos/blob/8040ff14c4463ae649821abea8b5244699672fc9/pkg/compact/compact.go#L231
to group only by external labels (from meta.json) and downsampling (from met.json as well) as it is safe to run concurrently such groups. Otherwise vertical compaction had to be enabled on target system and even then it's waste of compute someteims

B) We could do full concurrency as we want as long as we don't upload the last important file `meta.json`. With this target system is expected to not touch those blocks until delete period which is by defult 2 days. So we could do expensive/slow upload for blocks without meta.json and use option A for meta.json upload in time order.